src/global/project/manifest.rs

```diff
@@ -49,40 +42,180 @@ impl Manifest {
             contents
                 .parse::<DocumentMut>()
                 .map(|doc| (manifest, doc))
-                .map_err(ManifestError::from)
+                .map_err(TomlError::from)
         }) {
             Ok(result) => result,
             Err(e) => e.to_fancy(MANIFEST_DEFAULT_NAME, &contents)?,
         };
 
-        let source = ManifestSource(document);
         let manifest = Self {
             path: manifest_path.to_path_buf(),
-            contents,
-            document: source,
+            document: TomlManifest::new(document),
             parsed: manifest,
         };
 
         Ok(manifest)
     }
 
-    /// Adds an environment to the project.
-    pub fn add_environment(&mut self, _name: String) -> miette::Result<()> {
-        todo!()
+    pub fn add_exposed_mapping(
+        &mut self,
+        env_name: &EnvironmentName,
+        mapping: &Mapping,
+    ) -> miette::Result<()> {
+        self.document
+            .get_or_insert_nested_table(&format!("envs.{env_name}.exposed"))?
+            .insert(
+                &mapping.exposed_name.to_string(),
+                Item::Value(toml_edit::Value::from(mapping.executable_name.clone())),
+            );
+
+        tracing::debug!("Added exposed mapping {mapping} to toml document");
+        Ok(())
+    }
```

Comment:
This function brings this object in an inconsistent state where the TOML is updated but the parsed value is not. I consider this bad practice because after this operation is performed it is unclear whether a next operation will use the updated values or not.

We talked about it being ok to use this approach but I think in that case it should also be reflected in the types. For example by converting this type into a TomlManifest (dropping the parsed) and only adding this method on the `self.document` type.